ibis/tests/strategies.py

```diff
@@ -0,0 +1,187 @@
+    dt.Category, cardinality=st.integers(min_value=0), nullable=nullable
+)
+
+primitive_dtypes = st.one_of(
+    null_dtype,
+    boolean_dtype,
+    integer_dtypes,
+    floating_dtypes,
+    date_dtype,
+    time_dtype,
+)
+
+
+def array_dtypes(item_strategy=primitive_dtypes):
+    return st.builds(dt.Array, value_type=item_strategy, nullable=nullable)
+
+
+def set_dtypes(item_strategy=primitive_dtypes):
+    return st.builds(dt.Set, value_type=item_strategy, nullable=nullable)
+
+
+def map_dtypes(key_strategy=primitive_dtypes, value_strategy=primitive_dtypes):
+    return st.builds(
+        dt.Map, key_type=key_strategy, value_type=value_strategy, nullable=nullable
+    )
+
+
+@st.composite
+def struct_dtypes(
+    draw,
+    item_strategy=primitive_dtypes,
+    num_fields=st.integers(min_value=0, max_value=20),
+):
+    num_fields = draw(num_fields)
+    names = draw(st.lists(st.text(), min_size=num_fields, max_size=num_fields))
+    types = draw(st.lists(item_strategy, min_size=num_fields, max_size=num_fields))
+    return dt.Struct(names, types, nullable=draw(nullable))
+
+
+point_dtype = st.builds(dt.Point, nullable=nullable)
+linestring_dtype = st.builds(dt.LineString, nullable=nullable)
+polygon_dtype = st.builds(dt.Polygon, nullable=nullable)
+multipoint_dtype = st.builds(dt.MultiPoint, nullable=nullable)
+multilinestring_dtype = st.builds(dt.MultiLineString, nullable=nullable)
+multipolygon_dtype = st.builds(dt.MultiPolygon, nullable=nullable)
+geometry_dtype = st.builds(dt.Geometry, nullable=nullable)
+geography_dtype = st.builds(dt.Geography, nullable=nullable)
+geospatial_dtypes = st.one_of(
+    point_dtype,
+    linestring_dtype,
+    polygon_dtype,
+    multipoint_dtype,
+    multilinestring_dtype,
+    multipolygon_dtype,
+    geometry_dtype,
+    geography_dtype,
+)
+
+string_dtype = st.builds(dt.String, nullable=nullable)
+binary_dtype = st.builds(dt.Binary, nullable=nullable)
+json_dtype = st.builds(dt.JSON, nullable=nullable)
+jsonb_dtype = st.builds(dt.JSONB, nullable=nullable)
+inet_dtype = st.builds(dt.INET, nullable=nullable)
+macaddr_dtype = st.builds(dt.MACADDR, nullable=nullable)
+uuid_dtype = st.builds(dt.UUID, nullable=nullable)
+
+variadic_dtypes = st.one_of(
+    string_dtype,
+    binary_dtype,
+    json_dtype,
+    jsonb_dtype,
+    inet_dtype,
+    macaddr_dtype,
+    array_dtypes(),
+    set_dtypes(),
+    map_dtypes(),
+)
+
+all_dtypes = st.deferred(
+    lambda: (
+        primitive_dtypes
+        | interval_dtype
+        | category_dtype
+        | uuid_dtype
+        | geospatial_dtypes
+        | variadic_dtypes
+        | struct_dtypes()
+        | array_dtypes(all_dtypes)
+        | set_dtypes(all_dtypes)
+        | map_dtypes(all_dtypes, all_dtypes)
+        | struct_dtypes(all_dtypes)
+    )
+)
+
+
+# Strategies for generating schema
+
+
+@st.composite
+def schema(draw, item_strategy=primitive_dtypes, max_size=20):
+    num_fields = draw(st.integers(min_value=0, max_value=max_size))
```

Comment:
Should we still keep an empty schema as a valid construction?

Reply:
If an empty table is valid then an empty schema should be valid too.

Reply:
How about raising an exception on empty schema construction and disallowing them altogether?

Reply:
I wouldn't disallow empty schema construction, it is also allowed by pyarrow.